providers/kubeconfig/provider.go

```diff
@@ -0,0 +1,358 @@
+		p.client = mgr.GetLocalManager().GetClient()
+		if p.client == nil {
+			return fmt.Errorf("failed to get client from manager")
+		}
+	}
+
+	// Get the informer for secrets
+	secretInf, err := mgr.GetLocalManager().GetCache().GetInformer(ctx, &corev1.Secret{})
+	if err != nil {
+		return fmt.Errorf("failed to get secret informer: %w", err)
+	}
+
+	// Add event handlers for secrets
+	if _, err := secretInf.AddEventHandler(toolscache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			secret, ok := obj.(*corev1.Secret)
+			if !ok {
+				return false
+			}
+			// Only process secrets in our namespace with our label
+			return secret.Namespace == p.opts.Namespace &&
+				secret.Labels[p.opts.KubeconfigSecretLabel] == "true"
+		},
+		Handler: toolscache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				secret := obj.(*corev1.Secret)
+				log.Info("Processing new secret", "name", secret.Name)
+				if err := p.handleSecret(ctx, secret, mgr); err != nil {
+					log.Error(err, "Failed to handle secret", "name", secret.Name)
+				}
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				secret := newObj.(*corev1.Secret)
+				log.Info("Processing updated secret", "name", secret.Name)
+				if err := p.handleSecret(ctx, secret, mgr); err != nil {
+					log.Error(err, "Failed to handle secret", "name", secret.Name)
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				secret := obj.(*corev1.Secret)
+				log.Info("Processing deleted secret", "name", secret.Name)
+				p.handleSecretDelete(secret)
+			},
+		},
+	}); err != nil {
+		return fmt.Errorf("failed to add event handlers: %w", err)
+	}
+
+	// Block until context is done
+	<-ctx.Done()
+	log.Info("Context cancelled, exiting provider")
+	return ctx.Err()
+}
+
+// handleSecret processes a secret containing kubeconfig data
+func (p *Provider) handleSecret(ctx context.Context, secret *corev1.Secret, mgr mcmanager.Manager) error {
+	if secret == nil {
+		return fmt.Errorf("received nil secret")
+	}
+
+	// Extract name to use as cluster name
+	clusterName := secret.Name
+	log := p.log.WithValues("cluster", clusterName, "secret", fmt.Sprintf("%s/%s", secret.Namespace, secret.Name))
+
+	// Check if this secret has kubeconfig data
+	kubeconfigData, ok := secret.Data[p.opts.KubeconfigSecretKey]
+	if !ok {
+		log.Info("Secret does not contain kubeconfig data", "key", p.opts.KubeconfigSecretKey)
+		return nil
+	}
+
+	// Parse the kubeconfig
+	restConfig, err := clientcmd.RESTConfigFromKubeConfig(kubeconfigData)
+	if err != nil {
+		return fmt.Errorf("failed to parse kubeconfig: %w", err)
+	}
+
+	// Check if cluster exists and remove it if it does
+	p.lock.RLock()
+	_, clusterExists := p.clusters[clusterName]
+	p.lock.RUnlock()
+
+	if clusterExists {
+		log.Info("Cluster already exists, updating it")
+		if err := p.removeCluster(clusterName); err != nil {
+			return fmt.Errorf("failed to remove existing cluster: %w", err)
+		}
+	}
```

Comment:
I don't remember what hashing you were doing earlier, but looking at this code, won't every update to the kubeconfig secret coming down the informer pipeline cancel the engaged cluster and re-engage it? It probably makes sense to ensure we only do that if the kubeconfig actually changed.